File: ParProg-Testat-2/Aufgabe2/Java/BankAccount.java
package aufgabe2;

public class BankAccount {
	private int balance = 0;

	public void deposit(int amount) {
		balance += amount;
	}

	public boolean withdraw(int amount) {
		if (amount <= balance) {
			balance -= amount;
			return true;
		} else {
			return false;
		}
	}

	public int getBalance() {
		return balance;
	}
}

File: ParProg-Testat-2/ParProg_Testat-2_Fabian_Germann/Aufgabe_2/DotNet/BankAccount/BankAccount.cs
using System.Threading;

namespace BankAccount
{
    public class BankAccount
    {
        private int _balance = 0;

        public int Balance => Volatile.Read(ref _balance);

        public void Deposit(int amount)
        {
            Interlocked.Add(ref _balance, amount);
        }

        public bool Withdraw(int amount)
        {
            int initialValue, computedValue;
            do
            {
                initialValue = Balance;
                if (initialValue < amount)
                { 
                    return false;
                }

                computedValue = initialValue - amount;
            } while (initialValue != Interlocked.CompareExchange(ref _balance, computedValue, initialValue));

            return true;
        }
    }
}